src/object/cli_obj.c

```diff
@@ -3618,7 +3618,7 @@ obj_comp_cb(tse_task_t *task, void *data)
 	    DAOS_FAIL_CHECK(DAOS_DTX_NO_RETRY))
 		obj_auxi->io_retry = 0;
 
-	if (pm_stale || obj_auxi->io_retry)
+	if (!obj_auxi->no_retry && (pm_stale || obj_auxi->io_retry))
```

Comment:
I don't quite see why we need a special flag here, I suppose for both client and server stack, it should retry only for recoverable errors, for server shutdown, a non-recoverable error should be returned?

With current change, the rebuild will fail when pool leader changed? Is it acceptable?

Reply:
yes, leader change or pool map change will cause rebuild fail, but it will retry the rebuild anyway.

The original implementation will retry inside the client object stack, which might cause trouble if it want to abort the rebuild when it is in retry loop inside the object client stack.  There is no easy to notify the client stack to stop retry on server side.

So I choose to export all failure to migration, and let migration to check and retry.